esp/bindings/http/platform/httpservice.cpp

```diff
@@ -974,3 +875,552 @@ int CEspHttpServer::onGet()
+    IPropertyTree* sessionTree = domainSessions->queryBranch(xpath.str());
+    if (!sessionTree)
+    {
+        ESPLOG(LogMin, "Authentication failed: session:<%u> not found", sessionID);
+        if(authReq.isSoapPost) //from SOAP Test page
+            sendMessage("Session expired. Please close this page and login again.", "text/html; charset=UTF-8");
+        else
+            askUserLogin(authReq.authBinding, createHTTPSession(authReq, domainSessions, nullptr));
+
+        commitAndCloseRemoteConnection(conn);
+        return authFailed;
+    }
+
+    StringBuffer userName, password;
+    HTTPSessionState sessionState = (HTTPSessionState) sessionTree->getPropInt(PropSessionState);
+    if (sessionState == HTTPSS_new)
+    {//This request comes from a login page. Authenticate the user now.
+        if (!isEmptyString(_userName))
+            userName.set(_userName);
+        else
+            userName.set((authReq.requestParams) ? authReq.requestParams->queryProp("username") : nullptr);
+        if (!isEmptyString(_password))
+            password.set(_password);
+        else
+            password.set((authReq.requestParams) ? authReq.requestParams->queryProp("password") : nullptr);
+        if (userName.isEmpty() || password.isEmpty())
+        {
+            ESPLOG(LogMin, "Authentication failed: invalid credential for session:<%u>", sessionID);
+            handleAuthFailed(true, authReq, &sessionID);
+            commitAndCloseRemoteConnection(conn);
+            return authFailed;
+        }
+
+        authReq.ctx->setUserID(userName.str());
+        authReq.ctx->setPassword(password.str());
+        authReq.authBinding->populateRequest(m_request.get());
+        if (!authReq.authBinding->doAuth(authReq.ctx))
+        {
+            ESPLOG(LogMin, "Authentication failed for session:<%u>", sessionID);
+            handleAuthFailed(true, authReq, &sessionID);
+            commitAndCloseRemoteConnection(conn);
+            return authFailed;
+        }
+
+        // authenticate optional groups
+        authOptionalGroups(authReq);
+
+        sessionTree->setProp(PropSessionUserID, userName.str());
+        sessionTree->addPropInt(PropSessionState, HTTPSS_active);
+        authReq.ctx->addUserToken(sessionID);
+    }
+    else
+    {
+        authOptionalGroups(authReq);
+
+        //The UserID has to be set before the populateRequest() because the UserID is used to create the user object.
+        //After the user object is created, we may call addUserToken().
+        userName.set(sessionTree->queryProp(PropSessionUserID));
+        authReq.ctx->setUserID(userName.str());
+        authReq.authBinding->populateRequest(m_request.get());
+        authReq.ctx->addUserToken(sessionID);
+    }
+
+    ESPLOG(LogMax, "Authenticated for %s@%s", userName.str(), sessionTree->queryProp(PropSessionNetworkAddress));
+    postSessionAuth(authReq, sessionID, sessionState, timeNow, conn, sessionTree);
```

Comment:
I suspect postSessionAuth should get time again inside method, rather than be passed time established at head of this method.
Probably true for readAndCleanDomainSessions too, i.e. get time in that method too.